internal/testrunner/runners/system/runner.go

```diff
@@ -83,93 +87,201 @@ func (r *runner) TearDown() error {
+	if err != nil {
+		return result.withError(errors.Wrap(err, "reading service logs directory failed"))
+	}
+
+	var ctxt servicedeployer.ServiceContext
+	ctxt.Name = r.options.TestFolder.Package
+	ctxt.Logs.Folder.Local = serviceLogsDir
+	ctxt.Logs.Folder.Agent = serviceLogsAgentDir
+	testConfigs, err := newConfig(r.options.TestFolder.Path, ctxt)
+	if err != nil {
+		return result.withError(errors.Wrap(err, "unable to load system test configuration"))
+	}
+	// Run each configured test
+	for idx, testConfig := range testConfigs {
+		partial, err := r.runTest(testConfig, idx, ctxt)
+		results = append(results, partial...)
+		if err != nil {
+			return results, err
+		}
+		if err = r.TearDown(); err != nil {
+			return results, errors.Wrap(err, "failed to teardown runner")
+		}
+	}
+	return results, nil
+}
+
+func (r *runner) hasNumDocs(
+	dataStream string,
+	fieldsValidator *fields.Validator,
+	checker func(int) bool) func() (bool, error) {
+	return func() (bool, error) {
+		resp, err := r.options.ESClient.Search(
+			r.options.ESClient.Search.WithIndex(dataStream),
+			r.options.ESClient.Search.WithSort("@timestamp:asc"),
+			r.options.ESClient.Search.WithSize(elasticsearchQuerySize),
+		)
+		if err != nil {
+			return false, errors.Wrap(err, "could not search data stream")
+		}
+		defer resp.Body.Close()
+
+		var results struct {
+			Hits struct {
+				Total struct {
+					Value int
+				}
+				Hits []struct {
+					Source common.MapStr `json:"_source"`
+				}
+			}
 		}
 
-		if tcf, ok := err.(testrunner.ErrTestCaseFailed); ok {
-			tr.FailureMsg = tcf.Reason
-			tr.FailureDetails = tcf.Details
-			return []testrunner.TestResult{tr}, nil
+		if err := json.NewDecoder(resp.Body).Decode(&results); err != nil {
+			return false, errors.Wrap(err, "could not decode search results response")
 		}
 
-		tr.ErrorMsg = err.Error()
-		return []testrunner.TestResult{tr}, err
+		numHits := results.Hits.Total.Value
+		logger.Debugf("found %d hits in %s data stream", numHits, dataStream)
+		if !checker(numHits) {
+			return false, nil
+		}
+
+		var multiErr multierror.Error
+		for _, hit := range results.Hits.Hits {
+			if message, err := hit.Source.GetValue("error.message"); err != common.ErrKeyNotFound {
+				multiErr = append(multiErr, errors.New(message.(string)))
+				continue
+			}
+
+			errs := fieldsValidator.ValidateDocumentMap(hit.Source)
+			if errs != nil {
+				multiErr = append(multiErr, errs...)
+				continue
+			}
+		}
+
+		if len(multiErr) > 0 {
+			multiErr = multiErr.Unique()
+			return false, testrunner.ErrTestCaseFailed{
+				Reason:  fmt.Sprintf("one or more errors found in documents stored in %s data stream", dataStream),
+				Details: multiErr.Error(),
+			}
+		}
+		return true, nil
+	}
+}
+
+func (r *runner) runTest(config testConfig, cfgIdx int, ctxt servicedeployer.ServiceContext) ([]testrunner.TestResult, error) {
+	// Determine test name
+	name := config.Name
+	if name == "" {
+		name = config.Input
```

Comment:
I understand that the use case for this change is to test with different inputs but I think your solution is more generic than that. So I would not fallback to `config.Input` if `config.Name` is not defined. I think we can require developers to specify `config.Name` always, so if it's empty I think we should error out.